Correspondence Renderer

andreas-schneider opened the pull request:
The CorrespondenceMoMoRenderer allows to map points on the image with pionts on the mesh.
The renderCorrespondenceImage function returns a PixelImage containing TriangleFragments, which contains the TriangleId and the BarycentricCoordinates specifying the exact location on a TriangleMesh.

- separates rasterization and shading
- allows to reuse previously rasterized images for different applications such as rendering of depth maps or compute landmarks visibility

This PR contains:
- CorrespondenceMoMoRenderer
- Landmarks visibility calculation

Commit:
rendering of correspondence image with caching and landmarks visibility calculation.

Commit:
update sbt version.

BernhardEgger commented:
Thanks for that PR - it is a great new functionality. Please update the readme:
- add yourself to the list of contributors
- it would be nice to also add example correspondence renderings to not hide this new functionality

Commit:
Depth and color coded correspondence image rendering, choose rendering methods in config file, updated README.

Commit:
Added depth and correspondence image examples to README.

Commit:
.

Commit:
.

andreas-schneider commented:
Thank you for the suggestions!

I have now updated the PR with the following contents:

- Depth image rendering.
- Color coded correspondence rendering.
- Active rendering methods can be selected in the config file.
- Multiple rendering methods per sample.
- Updated readme with example images.

It is relatively easy to add new rendering methods.

Please have a look at it again.
Thanks

BernhardEgger commented:
Thanks a lot - looks now much better :)
Is sbt assembly still working after the change of sbt version?
I think Adam should merge - perhaps this nice contribution should even end in a new release?

Commit:
.

andreas-schneider commented:
Yes I just tried it. The tests are passing and I can run the synthesis from the jar file.
Yes I think so too :)

BernhardEgger commented:
👍

Andreas-Forster commented on src/main/scala/faces/apps/RandomFaces.scala:
```diff
@@ -101,27 +102,45 @@ object RandomFaces extends App {
+
         // render image with or without background
-        val img = if(bg) {
+        /*val img = if(bg) {
```
If this code is no longer used it should be deleted.

andreas-schneider replied:
yes

Andreas-Forster commented on src/main/scala/faces/settings/FacesSettingsJsonFormatV1.scala:
```diff
@@ -27,6 +28,8 @@ object RandomFacesSettingsJsonFormatV1 {
   import FacesSettingsJsonFormatV1._
   import scalismo.faces.io.renderparameters.RenderParameterJSONFormatV4._
 
+  implicit val ReneringMethodsFormat = RenderingMethodsJsonVormatV1.RenderingMethodsFormat
```
Typo: Vormat -> Format

andreas-schneider replied:
yes

Andreas-Forster commented on src/main/scala/faces/settings/FacesSettingsJsonFormatV1.scala:
```diff
@@ -179,6 +185,8 @@ object ControlledFacesSettingsJsonFormatV1 {
     }
   }
 
+  implicit val RenderingMethdosFormat = RenderingMethodsJsonVormatV1.RenderingMethodsFormat
```
Typo Methdos -> Methods

andreas-schneider replied:
yes

Andreas-Forster commented on src/main/scala/faces/settings/FacesSettingsJsonFormatV1.scala:
```diff
@@ -271,6 +282,30 @@ object ControlledFacesSettingsJsonFormatV1 {
 
 }
 
+object RenderingMethodsJsonVormatV1 {
```
Is it nescessary to have an own object? Or could this go into object FacesSettingsJsonFormatV1?

andreas-schneider replied:
Yes it could, I put it outside because it is used by the RandomFacesSettingsJsonFormatV1 and the FacesSettingsJsonFormatV1. Do you think I should put it into one of the two?

Andreas-Forster commented on src/main/scala/faces/utils/CorrespondenceMoMoRenderer.scala:
```diff
@@ -0,0 +1,188 @@
+import scala.collection.immutable
+
+/** parametric renderer for a Morphable Model, implements all useful Parameteric*Renderer interfaces */
+class CorrespondenceMoMoRenderer(val model: MoMo, val clearColor: RGBA)
```
Would it make sense to move this somewhen into scalismo-faces?

andreas-schneider replied:
Yes, I think we can do this after this PR.

Andreas-Forster commented on src/main/scala/faces/utils/CorrespondenceMoMoRenderer.scala:
```diff
@@ -0,0 +1,188 @@
+
+/** parametric renderer for a Morphable Model, implements all useful Parameteric*Renderer interfaces */
+class CorrespondenceMoMoRenderer(val model: MoMo, val clearColor: RGBA)
+  extends ParametricImageRenderer[RGBA]
```
Could we inherit from MoMoRenderer instead? Some functions are there already implemented as e.g. padCoefficients?

andreas-schneider replied:
Yes, I just committed the change.

Andreas-Forster reviewed (commented):
Thanks for this additional functionality. I have just a few comments that I see not critical for merging.

Commit:
Inherit from MoMoRenderer and small fixes.
